fuel-metrics/src/txpool_metrics.rs

```diff
@@ -0,0 +1,59 @@
+use lazy_static::lazy_static;
+pub use prometheus_client::metrics::histogram::Histogram;
+use prometheus_client::registry::Registry;
+use std::{
+    boxed::Box,
+    default::Default,
+    sync::RwLock,
+};
+
+pub struct TxPoolMetrics {
+    pub registry: Registry,
+    pub gas_price_histogram: Histogram,
+    pub tx_size_histogram: Histogram,
+}
+
+impl Default for TxPoolMetrics {
+    fn default() -> Self {
+        let registry = Registry::default();
+
+        let gas_prices = Vec::new();
+
+        let gas_price_histogram = Histogram::new(gas_prices.into_iter());
+
+        let tx_sizes = Vec::new();
+
+        let tx_size_histogram = Histogram::new(tx_sizes.into_iter());
+
+        Self {
+            registry,
+            gas_price_histogram,
+            tx_size_histogram,
+        }
+    }
+}
+
+pub fn init(mut metrics: TxPoolMetrics) -> TxPoolMetrics {
+    metrics.registry.register(
+        "Tx_Gas_Price_Histogram",
+        "A Histogram keeping track of all gas prices for each tx in the mempool",
+        Box::new(metrics.gas_price_histogram.clone()),
+    );
+
+    metrics.registry.register(
+        "Tx_Size_Histogram",
+        "A Histogram keeping track of the size of txs",
+        Box::new(metrics.tx_size_histogram.clone()),
+    );
+
+    metrics
+}
+
+lazy_static! {
+    pub static ref TXPOOL_METRICS: RwLock<TxPoolMetrics> = {
+        let registry = TxPoolMetrics::default();
+        let metrics = init(registry);
+
+        RwLock::new(metrics)
+    };
+}
```

Comment:
Global locks are almost always a bad idea. Is there a way we can make this not a global?

Reply:
I'd suggest just passing around the pool as:
```rust
struct TxPoolMetrics(Arc<RwLock<TxPoolMetricsInner>>);
```
Also avoid leaking the lock out to anything that uses the `TxPoolMetrics` and instead use methods that can keep the lock internal.

Reply:
Also given that the majority will be writing it's probably faster to use a `Mutex` then a `RwLock`

Reply:
Yeah it had initially started as just a way to prevent threads from both registering at the same time so I used a `RwLock`, would a `Mutex` actually end up being better if the lock is just moved down to the Registry (since registration occurs once, but a reading on `/metrics` occurs often?) It is unfortunately necessary to have metrics in this weird global pattern, since plumbing them through in each instance pollutes a lot of function parameters, and becomes even more difficult with feature flags

Reply:
If it's just about initializing once then maybe something like a `Once` could work and then you don't need the lock? I honestly think the plumbing is worth it. Globals cause so many issues. The sway repo is having a really hard time with them currently.

Reply:
I hadn't thought of `Once`, pinging @Voxelot since we also weighed the cost of global variables. I think it'd be interesting, but there is the cost that the counters need to be in scope of whatever they are measuring + the gql endpoint which can sometimes be a bit awkward. Counter's shouldn't need a lock though so using `Once` on Registries to attach counters could end up working out

Reply:
Once sounds interesting, I just don't have much exp with it yet.

As far as the globals... It's really egregious to pass these metrics around to where they are needed. For example in fuel-p2p, it requires plumbing through 3+ different p2p tasks / service objects before it gets to where it's needed.

I agree that globals are generally bad, but another way to look at it is to compare updating metrics to emitting log events. It needs to be done everywhere and in many places, and if a logging library didn't use thread-locals and required every function that wanted to log something to pass in a logger context I doubt anyone would use that library.

Reply:
I can see your point but in that case we should be leveraging something like `tracing` and just write a subscriber to our backend. They have done all the work to make a global safe and fast for this sort of thing. If this doesn't actually require a lock then it might be ok but even things like incorrectly setup atomics can greatly slow down an application if used everywhere.

Reply:
`tracing` isn't able to do metrics. There may be some crates like opentelemetry which attempt to do something similar, but it's not well maintained or up-to-date with the best prometheus libs.

Reply:
tracing can have any subscriber you want so it's very easy to create a metrics backend

Reply:
If we really need this to be a global then lets get rid of the locks

Reply:
Also keep in mind this will be global across all tests or multiple fuel-core instance running under the same process

Reply:
Locks are there for protection of the registry during registration and with gossipsub since mutable access is required during registration

Reply:
Writing the metrics as a tracing subscriber sounds interesting 🤔 One issue is that some of our deps like libp2p require direct access to a prometheus registry, so that may be more difficult to fit into the tracing model.

I'd prefer to have at least rudimentary telemetry in the near term even if that involves globals, while we come up with a more ideal solution.

Reply:
Talked with @freesig and I am going ot move the Registries into a `Once` cell, then expose that through global functions in `fuel-metrics`. `Once_cell` should be able to support the lazy initialization too of attaching counters to the registries, and since Counters are safe (Underlying type is in an `Arc`) we don't need a lock on those